File: 318.py
from typing import DefaultDict, List
from functools import reduce
from itertools import product

class Solution:
    def maxProduct(self, words: List[str]) -> int:
        masksHash = DefaultDict(int)
        for word in words:
            mask = reduce(lambda x, y : x | 1 << (ord(y) - ord('a')), word, 0)
            masksHash[mask] = max(masksHash[mask], len(word))   
        return max((masksHash[x] * masksHash[y] for x, y in product(masksHash, repeat=2) if x & y == 0), default=0)

File: go/318.go
package main

func maxProduct(words []string) (ans int) {
	n := len(words)
	masks := make([]int, n)
	for i, word := range words {
		mask := 0
		for _, ch := range word {
			mask |= 1 << (ch - 'a')
		}
		masks[i] = mask
	}
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			if masks[i]&masks[j] == 0 {
				ans = max(len(words[i])*len(words[j]), ans)
			}
		}
	}
	return
}
